### PR TITLE
[gpu] Change the implementation of GetFusionRoots to a simpler one

### PR DESCRIPTION
**Pull request by shraiysh**
Following the tuple -> get-tuple-element sequence for getting roots and remove allowing a duplicate root in case it is encountered twice in a row in the recursive search.

### CODE REVIEW

**Comment by shraiysh** [comment]
This check causes the test added in multi_output_fusion_test.cc to fail.

File: xla/service/gpu/gpu_fusible.cc [hunk 1]
```diff
@@ -885,12 +885,6 @@ static void GetFusionRootsRec(const HloInstruction* root,
       GetFusionRootsRec(root->operand(i), out);
     }
   } else {
-    if (!out.empty() && out.back() == root) {
-      return;
-    }
-    CHECK(!absl::c_linear_search(out, root))
```
Comment (akuegel) [inline xla/service/gpu/gpu_fusible.cc:891]: I think the check still makes sense, but in line 881 we assume that if we have a GetTupleElement op, it has as operand a Tuple op (which is not true in case of DynamicAddressFusion). So I think a better fix might be to change line 881 to ```if (root->opcode() == HloOpcode::kGetTupleElement && root->operand(0)->opcode() == HloOpcode::kTuple)```

See also https://github.com/openxla/xla/commit/2204f157da667491c1baa4f2f37a883cfc73068e for a similar recent fix.

Comment (shraiysh) [reply xla/service/gpu/gpu_fusible.cc:891]: Does this look okay? I removed the check from line 888 because it makes sense to not allow returning same values, but why we should be allowing them in adjacent elements in tuple is not clear to me.

Comment (akuegel) [reply xla/service/gpu/gpu_fusible.cc:891]: Yes, looks ok to me. I also don't understand why we try to allow a duplicate root in case it is encountered twice in a row in the recursive search. Seems arbitrary to me, and we should also check fail in that case.

**Review by akuegel — APPROVED** [review]
Thank you :)

File: xla/service/gpu/gpu_fusible.cc [hunk 2]
```diff
@@ -878,16 +878,16 @@ size_t GetOutputSizeOfFusible(const HloInstruction& instr) {
 // Recursive helper for GetFusionRoots below.
 static void GetFusionRootsRec(const HloInstruction* root,
                               std::vector<const HloInstruction*>& out) {
-  if (root->opcode() == HloOpcode::kGetTupleElement) {
+  if (root->opcode() == HloOpcode::kGetTupleElement &&
+      root->operand(0)->opcode() == HloOpcode::kTuple) {
```
Comment (akuegel) [inline xla/service/gpu/gpu_fusible.cc:882]: Your test case GpuFusibleTest.GetFusionRootsWithMakeTupleGTESequence shows that we will need an additional change to make this work correctly. We should only access the part of the tuple that the get-tuple-element refers to. So:

```
return GetFusionRootsRec(root->operand(0)->operand(root->tuple_index()), out);
```

File: xla/tests/build_defs.bzl [hunk 3]
```diff
@@ -111,7 +111,7 @@ def xla_test(
         if backend == "cpu":
             backend_deps = ["//xla/service:cpu_plugin"]
             backend_deps += ["//xla/tests:test_macros_cpu"]  # buildifier: disable=list-append
-        elif backend == "gpu":
+        elif backend == "gpu" or backend == "gpu_v100":
```
Comment (akuegel) [inline xla/tests/build_defs.bzl:114]: It looks to me that this belongs into a separate PR (I think I have even already seen it floating around)

File: xla/service/gpu/multi_output_fusion_test.cc [hunk 4]
```diff
@@ -2175,5 +2175,32 @@ ENTRY computation {
...
+    fusion {
+      p0 = s32[] parameter(0)
+      p1 = s32[32] parameter(1)
+      custom-call = (bf16[], s32[], u32[]) custom-call(p1), custom_call_target="my_custom_call"
+      get-tuple-element.0 = bf16[] get-tuple-element(custom-call), index=0
+      get-tuple-element.1 = s32[] get-tuple-element(custom-call), index=1
+      bitcast = s32[1] bitcast(get-tuple-element.1)
+      dynamic-update-slice = s32[32] dynamic-update-slice(p1, bitcast, p0)
+      get-tuple-element.2 = u32[] get-tuple-element(custom-call), index=2
+      ROOT tuple.30 = (bf16[], s32[32], u32[]) tuple(get-tuple-element.0 ,dynamic-update-slice, get-tuple-element.2)
```
Comment (akuegel) [inline xla/service/gpu/multi_output_fusion_test.cc:2191]: nit: put the comma directly after get-tuple-element.0

File: xla/service/gpu/gpu_fusible_test.cc [hunk 5]
```diff
@@ -1470,5 +1470,173 @@ ENTRY computation {
...
+      custom-call = (bf16[], s32[], u32[]) custom-call(p1), custom_call_target="my_custom_call"
+      get-tuple-element.0 = bf16[] get-tuple-element(custom-call), index=0
+      get-tuple-element.1 = s32[] get-tuple-element(custom-call), index=1
+      bitcast = s32[1] bitcast(get-tuple-element.1)
+      dynamic-update-slice = s32[32] dynamic-update-slice(p1, bitcast, p0)
+      get-tuple-element.2 = u32[] get-tuple-element(custom-call), index=2
+      tuple = (bf16[], s32[32], u32[]) tuple(get-tuple-element.0 ,dynamic-update-slice, get-tuple-element.2)
+      get-tuple-element.3 = bf16[] get-tuple-element(tuple), index=0
+      get-tuple-element.4 = u32[] get-tuple-element(tuple), index=2
+      ROOT tuple2 = (bf16[], s32[32], u32[]) tuple(get-tuple-element.3 ,dynamic-update-slice, get-tuple-element.4)
```
Comment (akuegel) [inline xla/service/gpu/gpu_fusible_test.cc:1617]: nit: put the comma directly after get-tuple-element.3

File: xla/service/gpu/gpu_fusible_test.cc [hunk 6]
```diff
@@ -1470,5 +1470,173 @@ ENTRY computation {
...
+    fusion {
+      p0 = s32[] parameter(0)
+      p1 = s32[32] parameter(1)
+      custom-call = (bf16[], s32[], u32[]) custom-call(p1), custom_call_target="my_custom_call"
+      get-tuple-element.0 = bf16[] get-tuple-element(custom-call), index=0
+      get-tuple-element.1 = s32[] get-tuple-element(custom-call), index=1
+      bitcast = s32[1] bitcast(get-tuple-element.1)
+      dynamic-update-slice = s32[32] dynamic-update-slice(p1, bitcast, p0)
+      get-tuple-element.2 = u32[] get-tuple-element(custom-call), index=2
+      tuple = (bf16[], s32[32], u32[]) tuple(get-tuple-element.0 ,dynamic-update-slice, get-tuple-element.2)
```
Comment (akuegel) [inline xla/service/gpu/gpu_fusible_test.cc:1614]: nit: put the comma directly after get-tuple-element.0

File: xla/service/gpu/gpu_fusible_test.cc [hunk 7]
```diff
@@ -1470,5 +1470,173 @@ ENTRY computation {
...
+    fusion {
+      p0 = s32[] parameter(0)
+      p1 = s32[32] parameter(1)
+      custom-call = (bf16[], s32[], u32[]) custom-call(p1), custom_call_target="my_custom_call"
+      get-tuple-element.0 = bf16[] get-tuple-element(custom-call), index=0
+      get-tuple-element.1 = s32[] get-tuple-element(custom-call), index=1
+      bitcast = s32[1] bitcast(get-tuple-element.1)
+      dynamic-update-slice = s32[32] dynamic-update-slice(p1, bitcast, p0)
+      get-tuple-element.2 = u32[] get-tuple-element(custom-call), index=2
+      ROOT tuple = (bf16[], s32[32], u32[]) tuple(get-tuple-element.0 ,dynamic-update-slice, get-tuple-element.2)
```
Comment (akuegel) [inline xla/service/gpu/gpu_fusible_test.cc:1578]: nit: put the comma directly after get-tuple-element.0

**Review by akuegel — CHANGES_REQUESTED** [review]
Please also update the PR description, I think it still refers to the change that the CHECK was removed.

**Comment by shraiysh** [comment]
I don't understand the things done by copybara here @akuegel. Was this merged in the previous state because of the approval, without waiting for nits to be addressed?

**Comment by akuegel** [comment]
> I don't understand the things done by copybara here @akuegel. Was this merged in the previous state because of the approval, without waiting for nits to be addressed?

No, these were just presubmits which also create a PR. You can ignore this.

**Review by akuegel — APPROVED** [review]
